src/mono/wasm/runtime/net6-legacy/hybrid-globalization.ts

```diff
@@ -84,6 +84,129 @@ export function pass_exception_details(ex: any, exceptionMessage: Int32Ptr){
+        return;
+    }
+    // pop is O(1) while shift is O(n) and reverse is O(n)
+    // it pays off to revert twice and pop more than shift multiple times
+    // unless we have < 2 empty chars at the beginning
+    if (segment.length === 1 || (segment.length > 1 && "".localeCompare(segment[1].segment, undefined) !== 0))
+    {
+        segment.shift();
+        return;
+    }
+    segment.reverse();
+    skip_zeroWidthChars_from_end(segment);
+    segment.reverse();
+}
+
+function skip_zeroWidthChars_from_end(segment: Intl.SegmentData[])
+{
+    while (segment.length !== 0 && "".localeCompare(segment[segment.length - 1].segment, undefined) === 0)
+    {
+        segment.pop();
+    }
+}
+
+export function mono_wasm_ends_with(exceptionMessage: Int32Ptr, culture: MonoStringRef, str1: number, str1Length: number, str2: number, str2Length: number, options: number, matchLengthPointer: number): number{
+    const cultureRoot = mono_wasm_new_external_root<MonoString>(culture);
+    try{
+        const cultureName = conv_string_root(cultureRoot);
+        const suffix = get_utf16_string(str2, str2Length);
+        // no need to look for an empty string
+        const result = "".localeCompare(suffix, undefined);
+        if (result === 0)
+            return 1; // true
+
+        const source = get_utf16_string(str1, str1Length);
+        const locale = cultureName ? cultureName : undefined;
+        const graphemesSource = segment_string_locale_sensitive(source, locale);
+        const graphemesSuffix = segment_string_locale_sensitive(suffix, locale);
+
+        // zero Width chars at the end do not matter in string comparison but are useful for calculating the match length
+        skip_zeroWidthChars_from_end(graphemesSource);
+        // zero Width chars do not matter in prefix's both ends
+        skip_zeroWidthChars_from_start(graphemesSuffix);
+        skip_zeroWidthChars_from_end(graphemesSuffix);
+
+        const casePicker = (options & 0x1f);
+        for (let i = 0; i < graphemesSuffix.length; i++)
+        {
+            const isEqual = compare_strings(graphemesSource[graphemesSource.length - 1 - i].segment, graphemesSuffix[graphemesSuffix.length - 1 - i].segment, locale, casePicker);
+            if (isEqual !== 0)
+                return 0; // false
+        }
+        const offset = graphemesSource.length - graphemesSuffix.length;
+        const firstCharIdx = graphemesSource[offset].index;
+        const matchLen = str1Length - firstCharIdx;
+        setU32(matchLengthPointer, matchLen);
+        return 1; // true
+    }
+    catch (ex: any) {
+        pass_exception_details(ex, exceptionMessage);
+        return -1;
+    }
+    finally {
+        cultureRoot.release();
+    }
+}
+
+export function segment_string_locale_sensitive(string: string, locale: string | undefined) : Intl.SegmentData[]
+{
+    const segmenter = new Intl.Segmenter(locale, { granularity: "grapheme" });
+    return Array.from(segmenter.segment(string));
+}
+
```

Comment:
`locale.split` on each call to `compare_strings` is probably expensive.

Reply:
I added performance tests and it seems it's not that bad.

| measurement | time ICU | time Hybrid |
|-:|-:|-:|
|           String, CompareInfo IsPrefix |     5.3330ms |     6.0471ms |
|           String, CompareInfo IsSuffix |     0.0091ms |     0.0093ms |
|                String, String IsPrefix |     5.5313ms |     6.0296ms |
|                String, String IsSuffix |     0.0094ms |     0.0102ms |

Reply:
Checked. There was an error connected with prefix testing but it does not change the range of ICU vs HG difference. Updated numbers:


| measurement | time ICU | time Hybrid |
|-:|-:|-:|
|           String, CompareInfo IsPrefix |     5.3665ms |     5.6897ms |
|           String, CompareInfo IsSuffix |     10.1670ms |     10.7293ms |
|                String, String IsPrefix |     5.4105ms |     6.4558ms |
|                String, String IsSuffix |     10.4519ms |     12.8267ms |